src/core/qgsreadwritecontext.h

```diff
@@ -41,8 +42,15 @@ class CORE_EXPORT QgsReadWriteContext
     //! Sets up path resolver for conversion between relative and absolute paths
     void setPathResolver( const QgsPathResolver &resolver ) { mPathResolver = resolver; }
 
+    //! append a message to the context
+    void pushMessage( Qgis::MessageLevel level, const QString &message ) {mMessages.append( qMakePair( level, message ) );}
+
+    //! return the stored messages and remove them
+    QList<QPair<Qgis::MessageLevel, QString>> takeMessages() {return mMessages; mMessages.clear();}
```

Comment:
I'm understanding that we were discouraging use of QPair in API now (since it's quite "opaque" and requires manual sip conversion code). Could we make this a small struct instead?

Reply:
fixed